### PR TITLE
*: add simple table structure check

### PR DESCRIPTION
**Pull request by IANTHEREAL**
* add table charset checking
* add table foreign key checking
* modify old checker code

### CODE REVIEW

**Comment by IANTHEREAL** [comment]
@csuzhangxc @dantin PTAL

File: pkg/check/table_structure.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+	var err error
+
+Loop:
+	for schema, tables := range c.tables {
+		if len(tables) == 0 {
+			tables, err = dbutil.GetTables(ctx, c.db, schema)
+			if err != nil {
+				markCheckError(r, err)
+				break Loop
```
Comment (dantin) [inline pkg/check/table_structure.go]: How about return `Result` here instead of break to outer loop.

File: pkg/check/table_structure.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+			}
+		}
+
+		for _, table := range tables {
+			statement, err := dbutil.GetCreateTableSQL(ctx, c.db, schema, table)
+			if err != nil {
+				markCheckError(r, err)
+				break Loop
```
Comment (dantin) [inline pkg/check/table_structure.go]: ditto

File: checker/main.go [hunk 3]
```diff
@@ -70,23 +53,32 @@ func main() {
...
+
+	schema := flag.Args()[0]
+	tables := flag.Args()[1:]
+	checkTables(schema, tables)
+	log.Infof("complete checking!!")
+}
+
+func checkTables(schema string, tables []string) {
+	db, err := dbutil.CreateDB(dbutil.DBConfig{
+		User:     *username,
```
Comment (csuzhangxc) [inline checker/main.go]: how about rename `dbutil.CreateDB` to `dbutil.OpenDB`, match `dbutil.CloseDB`

Comment (IANTHEREAL) [reply checker/main.go]: yep, it's better

File: pkg/check/table_structure.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+		db:     db,
+		tables: tables,
+	}
+}
+
+// Check implements Checker interface
+func (c *TablesChecker) Check(ctx context.Context) *Result {
+	r := &Result{
+		Name:  c.Name(),
+		Desc:  "table structure compatibility check",
```
Comment (csuzhangxc) [inline pkg/check/table_structure.go:37]: "checks table structure compatibility"? similar to other checkers

File: pkg/check/table_structure.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+		State: StateSuccess,
+	}
```
Comment (csuzhangxc) [inline pkg/check/table_structure.go:39]: init with `StateSuccess`, but `markCheckError` will not set it to `StateFailure`.
so, if `dbutil.GetTable` or `dbutil.GetCreateTableSQL` returns error (`c.errors == 0`), `r.State` still be `StateSuccess`.
change init value, or do `c.errors++`

File: pkg/check/table_structure.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb-tools/pkg/dbutil"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/parser"
+	"github.com/pingcap/tidb/util/charset"
+)
+
+// TablesChecker checks table structures
+type TablesChecker struct {
```
Comment (csuzhangxc) [inline pkg/check/table_structure.go]: any detail or ref?

File: pkg/check/table_structure.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+	st, ok := stmt.(*ast.CreateTableStmt)
+	if !ok {
+		return errors.Errorf("Expect CreateTableStmt but got %T", stmt)
+	}
+	var err error
+
+	// check columns
+	for _, def := range st.Cols {
+		err = c.checkColumnDef(def)
+		if err != nil {
```
Comment (csuzhangxc) [inline pkg/check/table_structure.go:103]: `c.checkColumnDef` has no effect? why add it

File: pkg/check/table_structure.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	// check options
+	for _, opt := range st.Options {
+		err = c.checkTableOption(opt)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return errors.Trace(err)
```
Comment (csuzhangxc) [inline pkg/check/table_structure.go]: `err != nil` cases all returned before, so we need `return errors.Trace(err)` here?

File: pkg/check/table_structure.go (hunk 7 above)
Comment (IANTHEREAL) [reply pkg/check/table_structure.go:103]: leave a blank and fill it out later. If we check compatibility, we need to check definitions of column. constraints and table options in generally

File: pkg/check/table_structure.go (hunk 4 above)
Comment (IANTHEREAL) [reply pkg/check/table_structure.go:37]: not matter, leave it flexible.  Serious and lively

File: pkg/check/table_structure.go (hunk 8 above)
Comment (IANTHEREAL) [reply pkg/check/table_structure.go]: maybe habits of some people, errors.Trace(nil) = nil

**Comment by dantin** [comment]
LGTM

**Review by csuzhangxc — APPROVED** [review]
LGTM

**Pull request merged**